File: archive/java/com/darkevilsudoku/board/SudokuIndexes.java
package com.darkevilsudoku.board;

import java.util.ArrayList;
import java.util.List;

public final class SudokuIndexes {
	private static final int[][] rows;
	private static final int[][] columns;
	private static final int[][] squares;
	private static final int[][] related;
	private static final int[] indexToSquare;
	private static final int[] indexToCellIndexSquare;

	static {
		rows = new int[SudokuBoard.SEQUENCE_SIZE][];
		columns = new int[SudokuBoard.SEQUENCE_SIZE][];
		squares = new int[SudokuBoard.SEQUENCE_SIZE][];
		related = new int[SudokuBoard.BOARD_SIZE][];
		indexToSquare = new int[SudokuBoard.BOARD_SIZE];
		indexToCellIndexSquare = new int[SudokuBoard.BOARD_SIZE];

		for (int i = 0; i < SudokuBoard.SEQUENCE_SIZE; i++) {
			rows[i] = new int[SudokuBoard.SEQUENCE_SIZE];
			columns[i] = new int[SudokuBoard.SEQUENCE_SIZE];
			squares[i] = new int[SudokuBoard.SEQUENCE_SIZE];

			for (int j = 0; j < SudokuBoard.SEQUENCE_SIZE; j++) {
				// set row indexes
				// 0 .. 8; 9 .. 17; ...
				rows[i][j] = i * SudokuBoard.SEQUENCE_SIZE + j;

				// set column indexes
				// { 0, 9, 18 ...} {1, 10, 19 ...} ...
				columns[i][j] = i + j * SudokuBoard.SEQUENCE_SIZE;

				// set square indexes
				// { 0, 1, 2, 9, 10, 11, ...}
				int index = 0;
				index += (i / 3) * 27; // note that i / 3 gives integer value so
										// the expression cannot be replaced
										// with i * 9
				index += (j / 3) * 9;
				index += ((i % 3) * 3 + (j % 3));

				squares[i][j] = index;

				// set map between index to square
				indexToSquare[index] = i;
				indexToCellIndexSquare[index] = j;
			}
		}

		// get the related indexes for each cell
		for (int i = 0; i < SudokuBoard.BOARD_SIZE; i++) {
			List<Integer> relatedList = new ArrayList<Integer>(20);
			for (int relatedIndex : rows[i / 9]) {
				if (relatedIndex == i)
					continue;
				if (!relatedList.contains(relatedIndex))
					relatedList.add(relatedIndex);
			}
			for (int relatedIndex : columns[i % 9]) {
				if (relatedIndex == i)
					continue;
				if (!relatedList.contains(relatedIndex))
					relatedList.add(relatedIndex);
			}
			for (int relatedIndex : squares[squareFromIndex(i)]) {
				if (relatedIndex == i)
					continue;
				if (!relatedList.contains(relatedIndex))
					relatedList.add(relatedIndex);
			}

			if (relatedList.size() != 20)
				throw new RuntimeException();
			related[i] = new int[20];
			for (int j = 0; j < 20; j++)
				related[i][j] = relatedList.get(j);
		}
	}

	public static int rowFromIndex(int index) {
		return index / 9;
	}

	public static int columnFromIndex(int index) {
		return index % 9;
	}

	public static int squareFromIndex(int index) {
		return indexToSquare[index];
	}

	public static int cellIndexInSquareFromIndex(int index) {
		return indexToCellIndexSquare[index];
	}

	public static int indexFromCoordinates(int row, int col) {
		return row * 9 + col;
	}

	public static int indexFromSquare(int squareIndex, int cellIndex) {
		return squares[squareIndex][cellIndex];
	}

	// TODO replace with immutable
	public static int[] getRelated(int index) {
		return related[index];
	}

	// TODO replace with immutable
	public static int[] getRowIndexes(int row) {
		return rows[row];
	}

	// TODO replace with immutable
	public static int[] getColumnIndexes(int column) {
		return columns[column];
	}

	// TODO replace with immutable
	public static int[] getSquareIndexes(int square) {
		return squares[square];
	}
	
	private SudokuIndexes() {
	}
}

File: cpp/main/board/SudokuIndexes.cpp
#include "SudokuIndexes.h"
#include <memory>
#include <string>
#include <stdexcept>

using namespace std;

array<sequence, 9> initRowIndexes()
{
    array<sequence, 9> rows;
    for (int row = 0; row < 9; row++)
    {
        sequence indexes;
        for (int column = 0; column < 9; column++)
        {
            indexes.at(column) = indexFromCoordinates(row, column);
        }
        rows.at(row) = std::move(indexes);
    }
    return std::move(rows);
}

array<sequence, 9> initColumnIndexes()
{
    array<sequence, 9> columns;
    for (int column = 0; column < 9; column++)
    {
        sequence indexes;
        for (int row = 0; row < 9; row++)
        {
            indexes.at(row) = indexFromCoordinates(row, column);
        }
        columns.at(column) = std::move(indexes);
    }
    return std::move(columns);
}

array<sequence, 9> initSquareIndexes()
{
    array<sequence, 9> squares;
    for (int square = 0; square < 9; square++)
    {
        sequence indexes;
        for (int cell = 0; cell < 9; cell++)
        {
            indexes.at(cell) = indexFromSquare(square, cell);
        }
        squares.at(square) = std::move(indexes);
    }
    return std::move(squares);
}

array<int, 81> initBoardIndexes()
{
    array<int, 81> indexes;
    for (int i = 0; i < 81; i++)
    {
        indexes.at(i) = i;
    }
    return std::move(indexes);
}

array<related, 81> initIndexToRelated()
{
    array<related, 81> all;
    for (int i = 0; i < 81; i++)
    {
        int relatedIndex = 0;
        int row = rowFromIndex(i);
        int column = columnFromIndex(i);
        int square = squareFromIndex(i);

        for (int rowIndex : getRowIndexes(row))
        {
            if (rowIndex == i)
            {
                continue;
            }
            all.at(i).at(relatedIndex++) = rowIndex;
        }
        for (int columnIndex : getColumnIndexes(column))
        {
            if (columnIndex == i)
            {
                continue;
            }
            all.at(i).at(relatedIndex++) = columnIndex;
        }
        for (int squareIndex : getSquareIndexes(square))
        {
            if (row == rowFromIndex(squareIndex) ||
                column == columnFromIndex(squareIndex))
            {
                continue;
            }
            all.at(i).at(relatedIndex++) = squareIndex;
        }
        if (relatedIndex != 20)
        {
            throw logic_error(string("expected # of related indexes to be 20, got ") + to_string(relatedIndex));
        }
    }

    return std::move(all);
}

array<sequence, 9> rowIndexes = initRowIndexes();
array<sequence, 9> columnIndexes = initColumnIndexes();
array<sequence, 9> squareIndexes = initSquareIndexes();
array<int, 81> boardIndexes = initBoardIndexes();
array<related, 81> indexToRelated = initIndexToRelated();

const sequence &getRowIndexes(int row)
{
    return rowIndexes.at(row);
}

const sequence &getColumnIndexes(int column)
{
    return columnIndexes.at(column);
}

const sequence &getSquareIndexes(int square)
{
    return squareIndexes.at(square);
}

const std::array<int, 81> &getBoardIndexes()
{
    return boardIndexes;
}

const sequence &getSequenceIndexes()
{
    return rowIndexes.at(0);
}

const related &getRelated(int index)
{
    return indexToRelated.at(index);
}
